src/python/grpcio_tests/tests/unit/_host_test.py

```diff
@@ -0,0 +1,145 @@
+import weakref
+
+import grpc
+from grpc import _channel
+
+from tests.unit import test_common
+from tests.unit.framework.common import test_constants
+
+_HOSTNAME = 'testhost'
+
+_REQUEST = b'\x00\x00\x00'
+_RESPONSE = b'\x00\x00\x00'
+
+_UNARY_UNARY = '/test/UnaryUnary'
+_UNARY_STREAM = '/test/UnaryStream'
+_STREAM_UNARY = '/test/StreamUnary'
+_STREAM_STREAM = '/test/StreamStream'
+
+
+class _MethodHandler(grpc.RpcMethodHandler):
+
+    def __init__(self):
+        self.request_streaming = False
+        self.response_streaming = False
+        self.request_deserializer = None
+        self.response_serializer = None
+        self._host = None
+
+    def unary_unary(self, request, servicer_context):
+        self._host = servicer_context.host()
+        return _RESPONSE
+
+    def unary_stream(self, request, servicer_context):
+        self._host = servicer_context.host()
+        for _ in range(test_constants.STREAM_LENGTH):
+            yield _RESPONSE
+
+    def stream_unary(self, request_iterator, servicer_context):
+        self._host = servicer_context.host()
+        # TODO(issue:#6891) We should be able to remove this loop
+        for request in request_iterator:
+            pass
+        return _RESPONSE
+
+    def stream_stream(self, request_iterator, servicer_context):
+        self._host = servicer_context.host()
+        # TODO(issue:#6891) We should be able to remove this loop,
+        # and replace with return; yield
+        for request in request_iterator:
+            yield _RESPONSE
+
+    def host(self):
+        return self._host
+
+
+class _GenericHandler(grpc.GenericRpcHandler):
+
+    def __init__(self):
+        self._handler = _MethodHandler()
+
+    def service(self, handler_call_details):
+        if handler_call_details.method == _UNARY_UNARY:
+            self._handler.request_streaming = False
+            self._handler.response_streaming = False
+        elif handler_call_details.method == _UNARY_STREAM:
+            self._handler.request_streaming = False
+            self._handler.response_streaming = True
+        elif handler_call_details.method == _STREAM_UNARY:
+            self._handler.request_streaming = True
+            self._handler.response_streaming = False
+        elif handler_call_details.method == _STREAM_STREAM:
+            self._handler.request_streaming = True
+            self._handler.response_streaming = True
+        else:
+            return None
+
+        return self._handler
+
+    def host(self):
+        return self._handler.host()
+
+
+class HostTest(unittest.TestCase):
+
+    def setUp(self):
+        self._server = test_common.test_server()
+        self._handler = _GenericHandler()
+        self._server.add_generic_rpc_handlers((self._handler,))
+        self._port = self._server.add_insecure_port('[::]:0')
+        self._server.start()
+        self._channel = grpc.insecure_channel('localhost:%d' % self._port)
+
+    def tearDown(self):
+        self._server.stop(0)
```

Comment:
Pass `None` here? If the server fails to stop we'd like the test itself to fail, right?